src/vm/comutilnative.cpp

```diff
@@ -2624,14 +2626,210 @@ FCIMPL6(INT32, ManagedLoggingHelper::GetRegistryLoggingValues, CLR_BOOL* bLoggin
 }
 FCIMPLEND
 
-// Return true if the valuetype does not contain pointer and is tightly packed
+typedef BOOL(*GetMethodTablePropertyFunc)(MethodTable*);
+typedef void(*SetMethodTablePropertyFunc)(MethodTable*);
+
+static BOOL HasOverriddenMethodWorker(MethodTable* mt, MethodTable* classMT, WORD methodSlot)
+{
+    CONTRACTL{
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+        SO_TOLERANT;
+    } CONTRACTL_END;
+
+    PCODE actual = mt->GetRestoredSlot(methodSlot);
+    PCODE base = classMT->GetRestoredSlot(methodSlot);
+
+    if (actual == base)
+    {
+        return FALSE;
+    }
+
+    if (!classMT->IsZapped())
+    {
+        // If mscorlib is JITed, the slots can be patched and thus we need to compare the actual MethodDescs
+        // to detect match reliably
+        if (MethodTable::GetMethodDescForSlotAddress(actual) == MethodTable::GetMethodDescForSlotAddress(base))
+        {
+            return FALSE;
+        }
+    }
+
+    return TRUE;
+}
+
+static BOOL HasOverriddenMethod(MethodTable* mt,
+                                MethodTable* classMT,
+                                WORD methodSlot,
+                                GetMethodTablePropertyFunc hasCheckedOverridden,
+                                SetMethodTablePropertyFunc setHasCheckedOverridden,
+                                GetMethodTablePropertyFunc hasOverridden,
+                                SetMethodTablePropertyFunc setHasOverridden)
+{
+    LIMITED_METHOD_CONTRACT
+
+    if (hasCheckedOverridden(mt))
+    {
+        return hasOverridden(mt);
+    }
+
+    BOOL hasOverriddenMethod = FALSE;
+
+    // Do we have the overridden method in current Type?
+    if (HasOverriddenMethodWorker(mt, classMT, methodSlot))
+    {
+        hasOverriddenMethod = TRUE;
+    }
+    else
+    {
+        // Now we need to do a depth-first-search on the instance fields.
+        Stack<ApproxFieldDescIterator> fieldIterStack;
```

Comment:
Any particular reason we did dfs recursively? I think it's slower than do it iteratively.